### PR TITLE
[5.8] Optionally allow recursively merging configs

### PR DESCRIPTION
**Pull request by koenhoeijmakers**
Up until now it was impossible to recursively merge configs which had multiple depths, this PR adds the option to do so. (see: #16991, #21541)

The added benefit of this is that a package creator can now properly "group" config keys instead of being limited to a single depth, which prevents the creator from having to release a mayor version when a new config key is added.

It is disabled by default so that the current functionality is not changed.

Example:
```
$one = ['jobs' => ['first'  => 'Programmer']];
$two = ['jobs' => ['second' => 'Teacher']];
```

Expected result _with_ recursive merging
```
$config =  ['jobs' => ['first'  => 'Programmer', 'second' => 'Teacher']];
```

Expected result _without_ recursive merging
```
$config =  ['jobs' => ['second' => 'Teacher']];
```

P.S. This is also a resubmit of #23680 without the extra helper.

### CODE REVIEW

File: src/Illuminate/Support/ServiceProvider.php [hunk 1]
```diff
@@ -56,7 +63,41 @@ protected function mergeConfigFrom($path, $key)
...
+    }
+
+    /**
+     * Merges the 2 given configs together, and if supplied, it will do it recursively.
+     *
+     * @param  array  $original
+     * @param  array  $merging
+     * @return array
+     */
+    protected function mergeConfigs($original, $merging)
```
[the remark this answers is not in the document]
Comment (koenhoeijmakers) [reply src/Illuminate/Support/ServiceProvider.php]: Certainly

[the remark this answers is not in the document]
Comment (koenhoeijmakers) [reply src/Illuminate/Support/ServiceProvider.php]: I suppose i can create subclass of the provider, i'll go with that approach

[the remark this answers is not in the document]
Comment (TBlindaruk) [reply src/Illuminate/Support/ServiceProvider.php]: I think you can test just public method, where protected is called

[the remark this answers is not in the document]
Comment (koenhoeijmakers) [reply src/Illuminate/Support/ServiceProvider.php]: So i'd have to boot the service provider, and then get the config and check if it's merged correctly from there?

What is the addition to that over my solution? (totally open to it ofcourse)

**Comment by taylorotwell** [comment]
No plans on adding this right now.

**Comment by herpaderpaldent** [comment]
damn i just ran into this issue. Is there a particular reason not to merge this?

**Pull request closed without merging**